File: coincidencetest/fca.py
"""
This module computes "formal concepts" (biclusters in binary feature data) using a
complete or truncated recursive pairwise closure strategy. The "lectic order" method
is not implemented. See [1]_ for a practical introduction to formal concepts, and
see [2]_ for a thorough treatment of the theory.

References
----------
.. [1] Ganter, Bernhard, and Obiedkov, Sergei. Conceptual Exploration. Germany,
       Springer Berlin Heidelberg, 2016.
.. [2] Ganter, Bernhard, and Stumme, Gerd. Formal Concept Analysis: Foundations and
       Applications. Germany, Springer, 2005.
"""
import random
from itertools import combinations

import pandas as pd

from .log_formats import colorized_logger
logger = colorized_logger(__name__)

class ConceptLattice:
    def __init__(self, data, level_limit: int=None, max_recursion: int=None):
        self.data = data
        self.level_limit = level_limit
        self.max_recursion = max_recursion

        if (not self.level_limit is None) or (not self.max_recursion is None):
            logger.debug('Using level limit %s, maximum recursions %s', self.level_limit, self.max_recursion)

        self.closed_sets = []
        self.dual_sets = []
        for feature in self.data.columns:
            closed_set, dual_set = self.closure([feature])
            if not self.already_have(closed_set):
                self.closed_sets.append(closed_set)
                self.dual_sets.append(dual_set)

        self.computed_pairs = []

    def compute_concepts(self):
        level = 1
        while True:
            previous_number_sets = len(self.closed_sets)
            new_pairs_computed = self.do_pairwise_closures()
            new_number_sets = len(self.closed_sets)
            if self.level_limit:
                logger.debug('Completed level %s. %s pairs tried, %s new sets.',
                    level,
                    self.level_limit,
                    new_number_sets - previous_number_sets,
                )
            else:
                logger.debug('Completed level %s. All pairs tried, %s new sets.',
                    level,
                    new_number_sets - previous_number_sets,
                )
            if previous_number_sets == new_number_sets:
                logger.debug('No new pairs at level %s.', level)
                return
            if self.max_recursion and level == self.max_recursion:
                logger.debug('Completed maximum level %s.', level)
                return
            level += 1

    def do_pairwise_closures(self):
        all_pairs = [
            tuple(sorted(c))
            for c in combinations(range(len(self.closed_sets)), 2)
        ]
        index_range = list(set(all_pairs).difference(self.computed_pairs))
        if len(index_range) == 0:
            return False
        new_pairs_computed = False
        if self.level_limit is not None:
            new_pairs = random.sample(index_range, self.level_limit)
        else:
            new_pairs = index_range
        for index1, index2 in new_pairs:
            union = sorted(list(
                set(self.closed_sets[index1]).union(self.closed_sets[index2])
            ))
            closed_set, dual_set = self.closure(union)
            if not self.already_have(closed_set) and len(dual_set) != 0:
                self.closed_sets.append(closed_set)
                self.dual_sets.append(dual_set)
                new_pairs_computed = True
            self.computed_pairs.append((index1, index2))
        return new_pairs_computed

    def closure(self, input_set):
        N = len(input_set)
        samples_mask = self.data.loc[:, input_set].apply(lambda row: sum(row) == N, axis=1)
        samples = sorted(list(self.data.index[samples_mask]))
        M = len(samples)
        features_mask = self.data.loc[samples, :].apply(lambda col: sum(col) == M, axis=0)
        features = sorted(list(self.data.columns[features_mask]))
        return features, samples

    def already_have(self, input_set):
        return input_set in self.closed_sets

def find_concepts(data, level_limit: int=None, max_recursion: int=None):
    """
    Computes the closure of each pair of features, then the closure of each pair of
    the resulting closed sets, etc. If `level_limit` is not None, the number of new
    closures to compute at each recursion level is `level_limit`.

    Parameters
    ----------
    data : pandas.DataFrame
        Binary data matrix, with row and column names. Rows are samples and columns
        are features.
    level_limit: int
        Default None. Limit on the number of new closures to compute per level. These are
        randomized among all possible pairs.
    max_recursion: int
        Default None. Limit on the number of levels.
    """
    lattice = ConceptLattice(data, level_limit=level_limit, max_recursion=max_recursion)
    lattice.compute_concepts()
    return [lattice.closed_sets, lattice.dual_sets]

File: webapp/fca.js

function arrays_equal(a, b) {
    if (a.length !== b.length) return false;
    for (var i = 0; i < a.length; ++i) {
        if (a[i] !== b[i]) {
            return false;
        }
    }
    return true;
}

function compute_closure(set, data) {
    let N = set.length
    let samples = []
    for (let i = 0; i < data["samples"].length; i++) {
        accumulator = 0
        for (let j = 0; j < set.length; j++) {
            accumulator += data["samples"][i][set[j]]
        }
        if (accumulator == N) {
            samples.push(i)
        }
    }
    samples.sort()

    let M = samples.length
    let features = []
    for (let i = 0; i < data["features"].length; i++) {
        accumulator = 0
        for (let j = 0; j < samples.length; j++) {
            accumulator += data["features"][i][samples[j]]
        }
        if (accumulator == M) {
            features.push(i)
        }
    }
    features.sort()

    return {"closed set" : features, "dual set" : samples}
}

function already_have(set, sets) {
    for (I in sets) {
        other = sets[I]
        if (arrays_equal(set, other)) {
            return true
        }
    }
    return false
}

function get_all_pairs(list) {
    let results = []
    for (let i = 0; i < list.length - 1; i++) {
        for (let j = i + 1; j < list.length; j++) {
            results.push([list[i], list[j]])
        }
    }
    return results
}

function get_random_integer(min, max) {
  return Math.floor(Math.random() * (max - min + 1) ) + min;
}

function get_random_sample(list, size) {
    let sampled = []
    while (sampled.length < size) {
        sample = getRandomInteger(0, list.length - 1)
        if (sampled.includes(sample)) {
            continue
        } else {
            sampled.push(sample)
        }
    }
    return Array.from(sampled, (i) => list[i]).sort();
}

function do_pairwise_closures(closed_sets, dual_sets, computed_pairs, level_limit, data) {
    let range = Array(closed_sets.length).fill().map((x,i)=>i)
    let all_pairs = get_all_pairs(range)
    
    let a = new Set(all_pairs)
    let b = new Set(computed_pairs)
    let index_range = new Set([...a].filter(x => !b.has(x)));

    if (index_range.length == 0) {
        return
    }

    if ( !(level_limit == null) ) {
        new_pairs = get_random_sample(index_range, level_limit)
    }
    else {
        new_pairs = [...index_range]
    }

    for (I in new_pairs) {
        pair = new_pairs[I]

        index1 = pair[0]
        index2 = pair[1]

        a = new Set(closed_sets[index1]);
        b = new Set(closed_sets[index2]);
        union = new Set([...a, ...b]);
        union = [...union].sort()
        c = compute_closure(union, data)

        if ( !(c["dual set"].length == 0) ) {
            if ( !(already_have(c["closed set"], closed_sets)) ) {
                closed_sets.push(c["closed set"])
                dual_sets.push(c["dual set"])
            }
        }
        computed_pairs.push([index1, index2])
    }
}

function find_concepts(data, level_limit, max_recursion) {
    closed_sets = []
    dual_sets = []

    columns = Array(data['features'].length).fill().map((x,i)=>i)
    for (I in columns) {
        feature = columns[I]
        c = compute_closure([feature], data)
        if ( !(already_have(c["closed set"], closed_sets)) && !(c["dual set"].length == 0) ) {
            closed_sets.push(c["closed set"])
            dual_sets.push(c["dual set"])
        }
    }

    computed_pairs = []
    level = 1
    while (true) {
        previous_number_sets = closed_sets.length
        new_pairs_computed = do_pairwise_closures(closed_sets, dual_sets, computed_pairs, level_limit, data)
        new_number_sets = closed_sets.length
        if ( previous_number_sets == new_number_sets ) {
            break
        }
        if ( !(max_recursion == null) ) {
            if(level == max_recursion) {
                break
            }
        }
        level += 1
    }

    function get_named(indices, names) {
        named = []
        for (let k = 0; k < indices.length; k++) {
            named.push(names[indices[k]])
        }
        return named
    }

    named_signatures = Array.from(closed_sets, (s) => get_named(s, data["header"]))

    return [named_signatures, dual_sets]
}

